File: sources/com/qx/level1/io/webgl/shapes/WebGL_Appearance.js

/**
 * CAD_Engine
 * 
 */
function WebGL_Appearance(){

	// Material

	/** the default program for wire rendering */
	this.wireProgramId = "color2";

	/** default value for shape material */
	this.wireColor = [0.12, 0.12, 0.12, 1.0];

	/** the default program for wire rendering */
	this.surfaceProgramId = "standard";

	/** default value for shape material -> "Standard" Unity-style shading */
	this.surfaceGlossiness = 0.7;

	/** default value for shape material -> "Standard" Unity-style shading */
	this.surfaceRoughness = 0.5;

	/** default value for shape material -> Phong shading */
	this.surfaceShininess = 0.5;

	/** default value for shape material -> multi-purposes */
	this.surfaceSpecularColor = [0.12, 0.8, 0.8, 1.0];

	/** default value for shape material -> multi-purposes */
	this.surfaceDiffuseColor = [0.12, 0.8, 0.8, 1.0];

	/** default value for shape material -> multi-purposes */
	this.surfaceAmbientColor = [0.1, 0.1, 0.1, 1.0];

	/** the textures */
	this.surfaceTexture0 = null;
	this.surfaceTexture1 = null;
	this.surfaceTexture2 = null;
	this.surfaceTexture3 = null;

}


BOHR_TYPES.set(WEBGL_BOHR_PREFIX+201, WebGL_Appearance);


WebGL_Appearance.prototype = {

		parse : function(input){

			// RGBA reading function
			function readRGBA8(input){
				var color = new Array(4);
				color[0] = input.getUInt8()/255.0; // r
				color[1] = input.getUInt8()/255.0; // g
				color[2] = input.getUInt8()/255.0; // b
				color[3] = input.getUInt8()/255.0; // a
				return color;
			};

			// <material>
			this.wireProgram = input.getStringUTF8();
			this.wireColor = readRGBA8(input);

			this.surfaceProgram = input.getStringUTF8();
			this.surfaceGlossiness = input.getFloat32();
			this.surfaceRoughness = input.getFloat32();
			this.surfaceShininess = input.getFloat32();
			this.surfaceSpecularColor = readRGBA8(input);
			this.surfaceDiffuseColor = readRGBA8(input);
			this.surfaceAmbientColor = readRGBA8(input);

			let texPathname;
			if((texPathname = input.getStringUTF8())!=null){
				this.surfaceTexture0 = WebGL_textures.get(texPathname);
			}
			if((texPathname = input.getStringUTF8())!=null){
				this.surfaceTexture1 = WebGL_textures.get(texPathname);
			}
			if((texPathname = input.getStringUTF8())!=null){
				this.surfaceTexture2 = WebGL_textures.get(texPathname);
			}
			if((texPathname = input.getStringUTF8())!=null){
				this.surfaceTexture3 = WebGL_textures.get(texPathname);
			}
			// </material>
		},

		resolve : function(objects){
			// nothing to resolve
		},
		
		build : function(){
			// nothing to build
		}
};

File: sources/com/qx/level1/io/webgl/shapes/WebGL_Appearance.java
package com.qx.level1.io.webgl.shapes;

import java.io.IOException;

import com.qx.level0.io.bohr.BohrObject;
import com.qx.level0.io.bohr.BohrObjectPrototype;
import com.qx.level0.io.bohr.BohrOutflow;
import com.qx.level1.io.webgl.WebGL_RGBA8Color;

public class WebGL_Appearance implements BohrObject {
	
	public final static BohrObjectPrototype BOHR_PROTOTYPE = new BohrObjectPrototype();
	public @Override BohrObjectPrototype getBohrPrototype() { return BOHR_PROTOTYPE; }
	

	/**
	 * default program for wire rendering 
	 */
	public String wireProgram = "color2";

	/**
	 * default value for shape material
	 */
	public WebGL_RGBA8Color wireColor = new WebGL_RGBA8Color(76, 76, 76, 255);

	/**
	 * default program for surface rendering 
	 */
	public String surfaceProgram = "standard";

	/**
	 * default value for shape material
	 */
	public double surfaceGlossiness = 0.7;


	/**
	 * default value for shape material
	 */
	public double surfaceRoughness = 0.5;


	/**
	 * Phong exponent for Old-style rendering
	 */
	public double surfaceShininess = 2.5;


	/** 
	 * default value for shape material
	 */
	public WebGL_RGBA8Color surfaceSpecularColor = new WebGL_RGBA8Color(255, 255, 255, 255);


	/** 
	 * default value for shape material
	 */
	public WebGL_RGBA8Color surfaceDiffuseColor = new WebGL_RGBA8Color(255, 255, 255, 255);


	/** 
	 * default value for shape material
	 */
	public WebGL_RGBA8Color surfaceAmbientColor = new WebGL_RGBA8Color(255, 255, 255, 255);


	/**
	 * surface color texture
	 */
	public String surfaceTexture0 = null;

	/**
	 * surface color texture
	 */
	public String surfaceTexture1 = null;

	/**
	 * surface color texture
	 */
	public String surfaceTexture2 = null;

	/**
	 * surface color texture
	 */
	public String surfaceTexture3 = null;



	@Override
	public void composeBody(BohrOutflow output) throws IOException {

		// material

		// <wire>
		output.putString(wireProgram);
		wireColor.write(output);
		// </wire>

		// <surface>
		output.putString(surfaceProgram);
		output.putFloat32((float) surfaceGlossiness);
		output.putFloat32((float) surfaceRoughness);
		output.putFloat32((float) surfaceShininess);
		surfaceSpecularColor.write(output);
		surfaceDiffuseColor.write(output);
		surfaceAmbientColor.write(output);

		// textures
		output.putString(surfaceTexture0);
		output.putString(surfaceTexture1);
		output.putString(surfaceTexture2);
		output.putString(surfaceTexture3);
		// </surface>
	}

}
